### DISCUSSION TITLE
Auto Update and every detail required

### DISCUSSION THREAD

**Discussion by TheLimeGlass** [SkriptLang/Skript#4652] [General]
Auto Update for Skript and every detail that would be needed in order for this feature to actually be possible.

Configuration:
- Config.sk configuration option, disabled by default. Boolean toggle.
- The comment of the configuration MUST state that the server may restart multiple times and that there cannot be any script errors for auto update to trigger.
- Have an option to only restart server if there are no players on.
- Have an option to only auto update server at a specific time.

Setup:
- When found, download latest version and place in plugins/update/ExistingJarNameOfSkript.jar so it replaces existing. Maybe only if named Skript.jar?
- Save old Skript jar to plugins/Skript/AutoUpdate/ExistingJarNameOfSkript.jar
- Write the old version and new version in a configuration file in plugins/Skript/AutoUpdate/info.txt or yaml
- Cancel auto update if there are currently scripts that already error.
- Cancel auto update if this update already caused errors from past auto update attempt (written in info configuration, more info below)
- Restart the server.

Startup:
- If any errors happen due to the new update, record the errors in plugins/Skript/AutoUpdate/errors.txt and set boolean in auto update info configuration that it errored.
- If erroring, revert the Skript version back to the old jar and restart the server. Delete the failed attempt Skript jar.
- While the latest version on GitHub is this exact version that errored in the info configuration, remind operators/players (with the update notification permission) and console that an auto update failed due to the new update and suggest them to please review the changelogs of that version, as one or more of their scripts errored, they can view said errors at the plugins/Skript/AutoUpdate/errors.txt

Might need to have a separate plugin to ensure Skript actually loads, because if Skript doesn't load that could be critical for the server. Say the latest version doesn't support their version of Minecraft because we dropped it.

We should also create a txt file within the Skript jar that tells the auto updater how to deal with updating from one version to another. This way we can set custom systems to execute when updating, such as the way player variables are serialized to the database, we could have a dedicated class to rewrite them rather than it being baked into the actual serializer code that exists now. Or that the auto update version isn't supported on this MC version, so cancel auto update. Etc.

Making all of which in a package org.skriptlang.skript.updater and have them able to be used by SkriptAddons with a well engineered API. Also have a SkriptAutoUpdateStartEvent, SkriptAutoUpdateSuccessEvent and SkriptAutoUpdateFailedEvent event. In addition to the already existing update checker a SkriptUpdateFoundEvent.

**Comment by Ankoki** [comment]
> Say the latest version doesn't support their version of Minecraft because we dropped it.  
  
At the start of the onEnable it could check the servers current Minecraft version and only pass with supported versions, to stop errors like that happening, I love this idea though!
